src/core/ComposableElement.php

```diff
@@ -357,6 +357,9 @@ private static function isSpreadKey(string $key): bool {
         $value = $this->validateAttributeValue($attr, $value);
       }
     } else {
+      // Skip null values for special attributes instead of casting them to empty string,
+      // since null is used to conditionally remove the values
+      if ($value === null) return $this;
```

Comment:
I think you need to handle this at render time instead of making `setAttribute` a no-op. As far as I can tell, this doesn't work (and if it does, adding it to the test case to verify would be useful IMO):

```
$x = <div data-foo="derp" />;
$x->setAttribute('data-foo', null);
expect($x->toString())->toBeSame('<div></div>');
```

Reply:
You're right that that example does not work.

What if I just remove the string cast? That seems to handle all existing test cases as well as the snippet you posted, and `renderBaseAttrs()` should be implicitly casting to string for non-null values anyway with the `htmlspecialchars()` calls.

Reply:
I added this change and updated the PR